workflows/multiomics/full_pipeline/main.nf

```diff
@@ -59,83 +59,47 @@ workflow run_wf {
-    | filter{ it[2].modality == "vdj" }
-    // | vdj_singlesample
-    | toSortedList{ a, b -> b[0] <=> a[0] }
-    | filter { it.size() != 0 } // filter when event is empty
-    | pmap{ list -> 
-      new_data = ["id": list.collect{it[0]}, "input": list.collect{it[1]}]
-      ["combined_vdj", new_data] + list[0].drop(2)
-    }
-    // | vdj_multisample
-    | concat.run(renameKeys: [input_id: "id"])
-
-      vdj_b_ch = start_ch
-    | filter{ it[2].modality == "vdj_b" }
-    // | vdj_singlesample
-    | toSortedList{ a, b -> b[0] <=> a[0] }
-    | filter { it.size() != 0 } // filter when event is empty
-    | pmap{ list -> 
-      new_data = ["id": list.collect{it[0]}, "input": list.collect{it[1]}]
-      ["combined_vdj_b", new_data] + list[0].drop(2)
-    }
-    // | vdj_b_multisample
-    | concat.run(renameKeys: [input_id: "id"])
+    modality_processors = [
+      [id: "rna", singlesample: rna_singlesample, multisample: rna_multisample],
+      [id: "vdj_t", singlesample: null, multisample: null],
+      [id: "vdj_b", singlesample: null, multisample: null],
+      [id: "prot", singlesample: null, multisample: null],
+      [id: "atac", singlesample: null, multisample: null],
+    ]
 
+    mod_chs = modality_processors.collect{ modality_processor ->
+      // Select the files corresponding to the currently selected modality
+      mod_ch = start_ch
+        | filter{ it[2].modality == modality_processor.id }
+        | view {"start channel-$modality_processor.id: $it"}
 
-  vdj_t_ch = start_ch
-    | filter{ it[2].modality == "vdj_t" }
-    // | vdj_singlesample
-    | toSortedList{ a, b -> b[0] <=> a[0] }
-    | filter { it.size() != 0 } // filter when event is empty
-    | pmap{ list -> 
-      new_data = ["id": list.collect{it[0]}, "input": list.collect{it[1]}]
-      ["combined_vdj_t", new_data] + list[0].drop(2)
+      // Run the single-sample processing if defined
+      ss_ch = (modality_processor.singlesample ? \
+               mod_ch | modality_processor.singlesample : \
+               mod_ch)
+      
+      // Reformat arguments to serve to the multisample processing
+      input_ms_ch = ss_ch
+        | view { "single-sample-input-$modality_processor.id: $it" }
+        | toSortedList{ a, b -> b[0] <=> a[0] }
+        | filter { it.size() != 0 } // filter when event is empty
+        | map{ list -> 
+          new_data = ["sample_id": list.collect{it[0]}, "input": list.collect{it[1]}]
+          ["combined_$modality_processor.id", new_data] + list[0].drop(2)
+        }
+        | view { "input multichannel-$modality_processor.id: $it" }
+      
+      // Run the multisample processing if defined, otherwise just concatenate samples together
+      out_ch = (modality_processor.multisample? \
+                input_ms_ch | modality_processor.multisample : \
+                input_ms_ch | concat.run(key: "concat_" + modality_processor.id,
+                                         renameKeys: [input_id: "sample_id"]))
```

Comment:
IMO you won't find this kind of indentation outside Python-esque languages.

I would replace this with:

```groovy
      out_ch = (
        modality_processor.multisample ? \
          input_ms_ch | modality_processor.multisample : \
          input_ms_ch | concat.run(
            key: "concat_" + modality_processor.id,
            renameKeys: [input_id: "sample_id"]
          )
      )
```

Reply:
This also looks good, better compared to the earlier proposed solution!